File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if(head==null) return null;
        //step -1
        Node ptr=head;
        while(ptr!=null){
            Node temp=new Node(ptr.val);
            temp.next=ptr.next;
            ptr.next=temp;
            ptr=ptr.next.next;
        }
        //ab random pointer
        ptr=head;
        while(ptr!=null){
            if(ptr.random!=null)
             ptr.next.random=ptr.random.next;
            ptr=ptr.next.next;
        }
        //divide karo ab
        ptr=head;
        Node f=head.next.next;
        Node head1=head.next;
        while(ptr!=null){
            if(f==null){
                ptr.next.next=null;
                ptr.next=null;
                break;
            }
            ptr.next.next=f.next;
            ptr.next=f;
            ptr=f;
            f=ptr.next.next;
        }

        return head1;
    }
}

File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x, next=None, random=None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        mapping = {}
        ptr = head
        mapping[None] = None

        while ptr != None:
            temp = Node(ptr.val)
            mapping[ptr] = temp
            ptr = ptr.next
        
        ptr = head
        copyHead = None
        while ptr != None:
            copy = mapping[ptr]
            if copyHead == None:
                copyHead = copy
            copy.next = mapping[ptr.next]
            copy.random = mapping[ptr.random]
            ptr = ptr.next
        
        return copyHead

        